python/lsst/sphgeom/box.cc

```diff
@@ -161,9 +162,9 @@ PYBIND11_MODULE(box, mod) {
     cls.def("__repr__", [](Box const &self) {
         return py::str("Box({!r}, {!r})").format(self.getLon(), self.getLat());
     });
-    cls.def("__setstate__", [](Box &self, py::bytes bytes) {
-        new (&self) Box(*decode(bytes));
-    });
+    cls.def(py::pickle(
+            [](const Box &self) { return python::encode(self); },
+            [](py::bytes bytes) { return decode(bytes).release(); }));
```

Comment:
Are you sure the explicit release is needed?

Reply:
Good question, but I was hoping you could tell me! I copied the code from the existing wrappers for `decode` (e.g. in `box.cc`, `ellipse.cc`, etc.). However, I did try building without `.release()` in the pickle wrapper (but not in the decode wrapper) and I didn't see any difference -- the unit tests passed. I figured it was safer to go with what was there, but it might be code that was never needed, or was only needed for older versions of `pybind11`. What do you think?

Reply:
http://pybind11.readthedocs.io/en/stable/advanced/smart_ptrs.html this implies that the release is not needed (and perhaps not wanted).

Reply:
Based on more tests I see the following:
- If I remove `.release()` from `clone` or `decode` in `region.cc` I get segfaults
- If I remove `.release()` from `decode` in `box.cc` I also get segfaults
- If I remove `.release()` from the pickle wrapper in `box.cc` I do not get segfaults

Based on that, and the fact that I don't understand why `release` is needed and the pickle docs say that the first function should look like `__getstate__`, which is all `decode` is, as far as I can tell, I propose to leave the `.release()` in everywhere. If you understand more deeply what is going on and think I can safely remove `.release()` from just the pickle support then I'm happy to do that.

Reply:
I agree with the proposed solution. Will have to look into this later. Perhaps file a new mini ticket? The `unique_ptr -> shared_ptr` shouldn't matter.